src/test/java/org/opensearch/knn/integ/ExpandNestedDocsIT.java

```diff
@@ -0,0 +1,333 @@
+        addVectorWithTopLevelField(4, numberOfNestedFields, FIELD_NAME_PARKING, FIELD_VALUE_TRUE);
+        addVectorWithTopLevelField(5, numberOfNestedFields, FIELD_NAME_PARKING, FIELD_VALUE_TRUE);
+        deleteKnnDoc(INDEX_NAME, String.valueOf(1));
+        updateVectorWithTopLevelField(2, numberOfNestedFields, FIELD_NAME_PARKING, FIELD_VALUE_FALSE);
+
+        // Run
+        Float[] queryVector = createVector();
+        Response response = queryNestedFieldWithExpandNestedDocs(INDEX_NAME, 10, queryVector, FIELD_NAME_PARKING, FIELD_VALUE_TRUE);
+
+        // Verify
+        String entity = EntityUtils.toString(response.getEntity());
+        Multimap<String, Integer> docIdToOffsets = parseInnerHits(entity, FIELD_NAME_NESTED);
+        assertEquals(3, docIdToOffsets.keySet().size());
+        for (String key : docIdToOffsets.keySet()) {
+            assertEquals(numberOfNestedFields, docIdToOffsets.get(key).size());
+        }
+    }
+
+    @SneakyThrows
+    public void testExpandNestedDocs_whenFilteredOnNestedFieldDoc_thenReturnFilteredNestedDoc() {
+        int numberOfNestedFields = 2;
+        createKnnIndex(engine, mode, dimension, dataType);
+        addVectorWithNestedLevelField(1, numberOfNestedFields, FIELD_NAME_STORAGE, Arrays.asList(FIELD_VALUE_FALSE, FIELD_VALUE_FALSE));
+        addVectorWithNestedLevelField(2, numberOfNestedFields, FIELD_NAME_STORAGE, Arrays.asList(FIELD_VALUE_TRUE, FIELD_VALUE_TRUE));
+        addVectorWithNestedLevelField(3, numberOfNestedFields, FIELD_NAME_STORAGE, Arrays.asList(FIELD_VALUE_TRUE, FIELD_VALUE_TRUE));
+        addVectorWithNestedLevelField(4, numberOfNestedFields, FIELD_NAME_STORAGE, Arrays.asList(FIELD_VALUE_FALSE, FIELD_VALUE_TRUE));
+        addVectorWithNestedLevelField(5, numberOfNestedFields, FIELD_NAME_STORAGE, Arrays.asList(FIELD_VALUE_TRUE, FIELD_VALUE_FALSE));
+        deleteKnnDoc(INDEX_NAME, String.valueOf(1));
+        addVectorWithNestedLevelField(2, numberOfNestedFields, FIELD_NAME_STORAGE, Arrays.asList(FIELD_VALUE_FALSE, FIELD_VALUE_FALSE));
+
+        // Run
+        Float[] queryVector = createVector();
+        Response response = queryNestedFieldWithExpandNestedDocs(
+            INDEX_NAME,
+            10,
+            queryVector,
+            FIELD_NAME_NESTED + "." + FIELD_NAME_STORAGE,
+            FIELD_VALUE_TRUE
+        );
+
+        // Verify
+        String entity = EntityUtils.toString(response.getEntity());
+        Multimap<String, Integer> docIdToOffsets = parseInnerHits(entity, FIELD_NAME_NESTED);
+        assertEquals(3, docIdToOffsets.keySet().size());
+        assertEquals(2, docIdToOffsets.get(String.valueOf(3)).size());
+        assertEquals(1, docIdToOffsets.get(String.valueOf(4)).size());
+        assertEquals(1, docIdToOffsets.get(String.valueOf(5)).size());
+
+        assertTrue(docIdToOffsets.get(String.valueOf(4)).contains(1));
+        assertTrue(docIdToOffsets.get(String.valueOf(5)).contains(0));
+    }
+
+    private Float[] createVector() {
```

Comment:
Do we need to reimplement this?

Reply:
Why do you think this need to be reimplemented?

Reply:
I meant we have similar functions that already exist

Reply:
Could you tell me where we have similar function?
For my case, it needs to be data type agnostic which means the value should be byte range and should be able to handle the dimension for both binary and non binary case.